### PR TITLE
KAFKA-13090: Improve kraft snapshot integration test

### PR DESCRIPTION
**Pull request by jsancio**
Check and verify generated snapshots for the controllers and the
brokers.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: raft/src/main/java/org/apache/kafka/raft/ReplicatedLog.java [hunk 1]
```diff
@@ -280,11 +280,19 @@ default long truncateToEndOffset(OffsetAndEpoch endOffset) {
...
     /**
-     * Returns the latest snapshot id if one exists.
+     * Returns the latest readable snapshot if one exists.
      *
-     * @return an Optional snapshot id of the latest snashot if one exists, otherwise returns an
-     * empty Optional
+     * @return an Optional with the latest readable snapshot, if one exists, otherwise
+     *         returns an empty Optional
      */
+    Optional<RawSnapshotReader> latestSnapshot();
```
Comment (cmccabe) [inline raft/src/main/java/org/apache/kafka/raft/ReplicatedLog.java:288]: would "openLatestSnapshot" be a better name, given that the snapshot reader will need to be closed later?

File: raft/src/main/java/org/apache/kafka/snapshot/SnapshotReader.java [hunk 2]
```diff
@@ -84,7 +84,7 @@ public int lastContainedLogEpoch() {
      */
     public long lastContainedLogTimestamp() {
         if (!lastContainedLogTimestamp.isPresent()) {
-            // nextBatch is expected to be empty
+            // nextBatch is expected to be equal to Optional.empty() so just replace it
```
Comment (cmccabe) [inline raft/src/main/java/org/apache/kafka/snapshot/SnapshotReader.java]: Can we check this and throw an exception if it's not true? I don't think this method is called often so performance should not be an issue.

File: raft/src/test/java/org/apache/kafka/raft/MockLog.java [hunk 3]
```diff
@@ -489,6 +489,11 @@ public void initializeLeaderEpoch(int epoch) {
         return Optional.ofNullable(snapshots.get(snapshotId));
     }
 
+    @Override
+    public Optional<RawSnapshotReader> latestSnapshot() {
+        return latestSnapshotId().flatMap(this::readSnapshot);
```
Comment (cmccabe) [inline raft/src/test/java/org/apache/kafka/raft/MockLog.java:494]: I didn't realize they implemented flatMap on Java's Optional, interesting.

**Review by cmccabe — APPROVED** [review]
LGTM pending Jenkins. I left some comments

File: raft/src/main/java/org/apache/kafka/raft/ReplicatedLog.java (hunk 1 above)
Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/raft/ReplicatedLog.java:288]: `RawSnapshotReader` are not `Closeable` or `AutoCloseable`. The specific implementation `FileRawSnapshotReader` is `AutoCloseable`.

This was done to match Kafka ownership model. For example, kraft sends the `FileChannel` owned by `RawSnapshotReader` to the networking layer when replying to `FetchSnapshot` requests.

`FileRawSnapshotReader` are closed as part of the clean up process. 60 seconds (if I remember correctly) after `maybeClean` has determine that they should be deleted.

File: raft/src/main/java/org/apache/kafka/snapshot/SnapshotReader.java (hunk 2 above)
Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/snapshot/SnapshotReader.java]: Done.

**Review by jsancio — COMMENTED** [review]
Thanks for the review @cmccabe .

**Comment by cmccabe** [comment]
Committed and cherry-picked to 3.0

**Pull request closed without merging**